keps/sig-cluster-lifecycle/kubeadm/20190722-Advanced-configurations-with-kubeadm-(Kustomize).md

```diff
@@ -0,0 +1,292 @@
+
+kubeadm maintainers should take care of making differences cristal clear in release notes
+and feature announcement:
+
+- The config API is well suited for most common use cases and it will remain the
+  recommended way to define core settings for a cluster, while “advanced configurations”
+  are designed for less common use cases.
+- The config API implicitly defines the main cluster variants the kubeadm team is
+  committed to support and monitor in the Kubernetes test grid, while instead higher-level
+  tools/user are responsible for the security, conformance, consistency, of 
+  “advanced configurations” for control-plane or kubelet settings. 
+
+_To not provide expected flexibility_
+In order to provide guarantee about kubeadm respecting “advanced configurations” during
+init, join, upgrades or single-phase execution, it is necessary to define some trade-offs
+around _what_ can be customized and _how_.
+
+Even if the proposed solution is based on the user feedback/issues, the kubeadm maintainers
+want to be sure the implemented is providing the expected level of flexibility and, in
+order to ensure that, we will wait for at least one K8s release cycle for the users to provide 
+feedback before moving forward in graduating the feature. 
+
+Similarly, the kubeadm maintainers should work with [`etcdadm`](https://github.com/kubernetes-sigs/etcdadm)
+project and [`addons`](https://github.com/kubernetes-sigs/addon-operators) project
+to ensure a consistent approach across different components.
+
+## Design Details
+
+### Test Plan
+
+Add at least a new E2E periodic test exercising  “advanced configurations” during init,
+join and upgrades.
+
+### Graduation Criteria
+
+This proposal in its initial version covers only the creation of a new alpha features.
+Graduation criteria will be defined in the following iterations on this proposal and
+consider user feedback as well.
+
+### Upgrade / Downgrade Strategy
+
+As stated in goals, kubeadm will preserve “advanced configurations” during upgrades,
+and more specifically, it will re-apply patches after each upgrade.
+
+Downgrades are not supported by kubeadm, and not considered by this proposal.
+
+### Version Skew Strategy
+
+This proposal does not impact kubeadm compliance with official K8s version skew policy;
+higher-level tools/user are responsible for the security, conformance, consistency of
+“advanced configurations” that can impact or the aforementioned point.
+
+## Implementation History
+
+Tuesday, July 30, 2019
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+
+## Drawbacks
+
+Kubeadm already offers a way to implement cluster settings, that is the kubeadm API
+and the support for providing component configs. Adding a new feature for supporting
+“advanced configurations” can create confusion in the users.
+
+See risks and mitigations.
```

Comment:
overall, i have concerns about the UX and i think the users might face difficulties.

while kustomize is becoming the standard of applying customizations in the k8s space it seems we are fighting the sandbox we created. i think we should be answering the questions of "why are we limit the API?" and "what do the users want from it?".

with the kubeadm API we have sandboxed the user to something that is maintainable for us and now we are exposing a mechanic to modify the sandbox. this might be difficult to manage for the users and will introduce some complexity on our side too. perhaps the sandbox should have not existed in the first place and instead we should have used the method of "sane and tested defaults, but still possible to use this API as a Swiss army knife".

Reply:
TBH, I think that those points should be discussed in the kubeadm office hours. 
For the track, my personal opinion on the first two questions

> "why are we limit the API?"

Past experience in managing kubeadm IMO already proved that accepting any kind of knobs as a CLI flag or config flags is not sustainable...

> "what do the users want from it?"

It depends. Some users want
- simplicity, and kubeadm already support this with the CLI and a limited set of flags.
- some flexibility, but with the right abstraction from all the internal, and kubeadm already support this with the kubeadm component configuration.
- the possibility to tune all the knobs for static pod manifests and this KEP is proposing a possible escape path for those users.